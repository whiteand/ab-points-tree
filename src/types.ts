export interface ITree<Point extends number[]> {
  point: Point;
  ranges: Array<[number, number]>;
  left: ITree<Point> | null;
  right: ITree<Point> | null;
}
export interface IBuildConfig<Point extends number[]> {
  // amount of generated points per each generation step
  chunkSize?: number;
  // return true in order to chunk updates
  chunkPredicate?: (
    point: Point,
    index: number,
    tree: ITree<Point>,
    currentChunkSize: number,
  ) => boolean;
  initialTree?: ITree<Point>;
}
