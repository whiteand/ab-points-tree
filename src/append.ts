import { tree as getTree } from './tree';
import { ITree } from './types';

function _append<Point extends number[]>(
  tree: ITree<Point> | null,
  point: Point,
): ITree<Point> {
  if (!tree) {
    const ranges = point.map(value => [value, value] as [number, number]);
    return getTree(point, ranges, null, null);
  }

  const headX = tree.point[0];
  const pointX = point[0];
  const ranges = Array(point.length);

  if (headX === pointX) {
    ranges[0] = tree.ranges[0];

    for (let i = 1; i < point.length; i++) {
      const leftRange = tree.left ? tree.left.ranges[i] : null;
      const rightRange = tree.right ? tree.right.ranges[i] : null;
      const y = point[i];
      let minY = y;
      let maxY = y;

      if (leftRange) {
        if (leftRange[0] < minY) {
          minY = leftRange[0];
        }
        if (leftRange[1] > maxY) {
          maxY = leftRange[1];
        }
      }

      if (rightRange) {
        if (rightRange[0] < minY) {
          minY = rightRange[0];
        }
        if (rightRange[1] > maxY) {
          maxY = rightRange[1];
        }
      }

      ranges[i] = [minY, maxY];
    }

    return getTree(point, ranges, tree.left, tree.right);
  }
  if (headX > pointX) {
    const newLeft = append(tree.left, point);

    for (let i = 0; i < point.length; i++) {
      let minY = tree.point[i];
      let maxY = tree.point[i];
      const leftRange = newLeft.ranges[i];
      const rightRange = tree.right ? tree.right.ranges[i] : null;
      if (leftRange[0] < minY) {
        minY = leftRange[0];
      }
      if (leftRange[1] > maxY) {
        maxY = leftRange[1];
      }
      if (rightRange) {
        if (rightRange[0] < minY) {
          minY = rightRange[0];
        }
        if (rightRange[1] > maxY) {
          maxY = rightRange[1];
        }
      }
      ranges[i] = [minY, maxY];
    }

    return getTree(tree.point, ranges, newLeft, tree.right);
  }
  const newRight = append(tree.right, point);

  for (let i = 0; i < point.length; i++) {
    let minY = tree.point[i];
    let maxY = tree.point[i];
    const leftRange = tree.left ? tree.left.ranges[i] : null;
    const rightRange = newRight.ranges[i];
    if (leftRange) {
      if (leftRange[0] < minY) {
        minY = leftRange[0];
      }
      if (leftRange[1] > maxY) {
        maxY = leftRange[1];
      }
    }
    if (rightRange[0] < minY) {
      minY = rightRange[0];
    }
    if (rightRange[1] > maxY) {
      maxY = rightRange[1];
    }
    ranges[i] = [minY, maxY];
  }

  return getTree(tree.point, ranges, tree.left, newRight);
}

export function append<Point extends number[]>(
  tree: ITree<Point> | null,
  ...points: Point[]
): ITree<Point> {
  let res = _append(tree, points[0]);
  for (let i = 1; i < points.length; i++) {
    const point = points[i];
    res = _append(res, point);
  }
  return res;
}
