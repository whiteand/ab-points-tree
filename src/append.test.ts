import { append } from './append';
import { ITree } from './types';

describe('append function', () => {
  test('append to empty tree 1-d point', () => {
    const d1Tree = append(null, [1]);
    expect(d1Tree).toMatchSnapshot();
  });
  test('append to empty tree 2-d point', () => {
    const d1Tree = append(null, [1, 2]);
    expect(d1Tree).toMatchSnapshot();
  });
  test('append to empty tree 3-d point', () => {
    const d1Tree = append(null, [1, 2, 3]);
    expect(d1Tree).toMatchSnapshot();
  });
  test('append instead of head without right and left', () => {
    const t = append(null, [1, 2]);
    const newT = append(t, [1, 3]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, without changing of ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 16],
        ranges: [
          [4, 4],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 36],
        ranges: [
          [6, 6],
          [36, 36],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 17]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, decreasing without changing of ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 36],
        ranges: [
          [4, 4],
          [36, 36],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 16],
        ranges: [
          [6, 6],
          [16, 16],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 17]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, with min change ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 16],
        ranges: [
          [4, 4],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 36],
        ranges: [
          [6, 6],
          [36, 36],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 10]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, with max change ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 16],
        ranges: [
          [4, 4],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 36],
        ranges: [
          [6, 6],
          [36, 36],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 40]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, decreasing, with min change ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 36],
        ranges: [
          [4, 4],
          [36, 36],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 16],
        ranges: [
          [6, 6],
          [16, 16],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 10]);
    expect(newT).toMatchSnapshot();
  });
  test('append instead of head with right and left, decreasing, with max change ranges', () => {
    const tree: ITree<[number, number]> = {
      left: {
        left: null,
        point: [4, 36],
        ranges: [
          [4, 4],
          [36, 36],
        ],
        right: null,
      },
      point: [5, 25],
      ranges: [
        [4, 6],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 16],
        ranges: [
          [6, 6],
          [16, 16],
        ],
        right: null,
      },
    };
    const newT = append(tree, [5, 40]);
    expect(newT).toMatchSnapshot();
  });
  test('append one leap', () => {
    const t = [5, 4].reduce(
      (tree: ITree<[number, number, number]> | null, x) =>
        append(tree, [x, x + 1, x * x] as [number, number, number]),
      null,
    );
    expect(t).toEqual({
      left: {
        left: null,
        point: [4, 5, 16],
        ranges: [
          [4, 4],
          [5, 5],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 6, 25],
      ranges: [
        [4, 5],
        [5, 6],
        [16, 25],
      ],
      right: null,
    });
  });
  test('append leaps', () => {
    const t = [5, 4, 6].reduce(
      (tree: ITree<[number, number, number]> | null, x) =>
        append(tree, [x, x + 1, x * x] as [number, number, number]),
      null,
    );
    expect(t).toEqual({
      left: {
        left: null,
        point: [4, 5, 16],
        ranges: [
          [4, 4],
          [5, 5],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 6, 25],
      ranges: [
        [4, 6],
        [5, 7],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 7, 36],
        ranges: [
          [6, 6],
          [7, 7],
          [36, 36],
        ],
        right: null,
      },
    });
  });
  test('append multiple points', () => {
    const t = append(
      null,
      ...[5, 4, 6].map(x => [x, x + 1, x * x] as [number, number, number]),
    );
    expect(t).toEqual({
      left: {
        left: null,
        point: [4, 5, 16],
        ranges: [
          [4, 4],
          [5, 5],
          [16, 16],
        ],
        right: null,
      },
      point: [5, 6, 25],
      ranges: [
        [4, 6],
        [5, 7],
        [16, 36],
      ],
      right: {
        left: null,
        point: [6, 7, 36],
        ranges: [
          [6, 6],
          [7, 7],
          [36, 36],
        ],
        right: null,
      },
    });
  });
});
