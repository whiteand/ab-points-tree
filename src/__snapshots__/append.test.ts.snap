// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`append function append instead of head with right and left, decreasing without changing of ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      36,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    17,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      16,
      36,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      16,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head with right and left, decreasing, with max change ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      36,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    40,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      16,
      40,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      16,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head with right and left, decreasing, with min change ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      36,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    10,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      10,
      36,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      16,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head with right and left, with max change ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      16,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    40,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      16,
      40,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      36,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head with right and left, with min change ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      16,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    10,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      10,
      36,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      36,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head with right and left, without changing of ranges 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      4,
      16,
    ],
    "ranges": Array [
      Array [
        4,
        4,
      ],
      Array [
        16,
        16,
      ],
    ],
    "right": null,
  },
  "point": Array [
    5,
    17,
  ],
  "ranges": Array [
    Array [
      4,
      6,
    ],
    Array [
      16,
      36,
    ],
  ],
  "right": Object {
    "left": null,
    "point": Array [
      6,
      36,
    ],
    "ranges": Array [
      Array [
        6,
        6,
      ],
      Array [
        36,
        36,
      ],
    ],
    "right": null,
  },
}
`;

exports[`append function append instead of head without right and left 1`] = `
Object {
  "left": null,
  "point": Array [
    1,
    3,
  ],
  "ranges": Array [
    Array [
      1,
      1,
    ],
    Array [
      3,
      3,
    ],
  ],
  "right": null,
}
`;

exports[`append function append to empty tree 1-d point 1`] = `
Object {
  "left": null,
  "point": Array [
    1,
  ],
  "ranges": Array [
    Array [
      1,
      1,
    ],
  ],
  "right": null,
}
`;

exports[`append function append to empty tree 2-d point 1`] = `
Object {
  "left": null,
  "point": Array [
    1,
    2,
  ],
  "ranges": Array [
    Array [
      1,
      1,
    ],
    Array [
      2,
      2,
    ],
  ],
  "right": null,
}
`;

exports[`append function append to empty tree 3-d point 1`] = `
Object {
  "left": null,
  "point": Array [
    1,
    2,
    3,
  ],
  "ranges": Array [
    Array [
      1,
      1,
    ],
    Array [
      2,
      2,
    ],
    Array [
      3,
      3,
    ],
  ],
  "right": null,
}
`;
